trie-db/src/lookup.rs

```diff
@@ -655,4 +668,175 @@ where
+							// This effectively returns the hash of the `Node::Branch` at the next
+							// iteration.
+							common_prefix_len == partial.len()
+						} else {
+							// Remainder of the provided key is longer than the extension slice,
+							// must advance the node iteration if and only if keys share
+							// a common prefix.
+							common_prefix_len == slice.len()
+						};
+
+						if start_matches {
+							// Empties the partial key if the extension slice is longer.
+							partial = partial.mid(common_prefix_len);
+							key_nibbles += common_prefix_len;
+							item
+						} else {
+							self.record(|| TrieAccess::NonExisting { full_key });
+
+							return Ok(None)
+						}
+					},
+					Node::Branch(children, value) =>
+						if partial.is_empty() {
+							if value.is_none() {
+								self.record(|| TrieAccess::NonExisting { full_key });
+							}
+
+							return Ok(Some(hash))
+						} else {
+							match children[partial.at(0) as usize] {
+								Some(x) => {
+									partial = partial.mid(1);
+									key_nibbles += 1;
+									x
+								},
+								None => {
+									self.record(|| TrieAccess::NonExisting { full_key });
+
+									return Ok(None)
+								},
+							}
+						},
+					Node::NibbledBranch(slice, children, value) => {
+						let common_prefix_len = partial.common_prefix(&slice);
+						// Not enough remainder key to continue the search.
+						if partial.len() < slice.len() {
+							self.record(|| TrieAccess::NonExisting { full_key });
+
+							// Branch slice starts with the remainder key, there's nothing to
+							// advance.
+							if common_prefix_len == partial.len() {
+								return Ok(Some(hash))
+							} else {
+								return Ok(None)
+							}
+						}
+
+						// Partial key is longer or equal than the branch slice.
+						// Ensure partial key starts with the branch slice.
+						if common_prefix_len != slice.len() {
+							self.record(|| TrieAccess::NonExisting { full_key });
+							return Ok(None)
+						}
+
+						if partial.len() == slice.len() {
+							if value.is_none() {
+								self.record(|| TrieAccess::NonExisting { full_key });
+							}
+
+							return Ok(Some(hash))
+						} else {
+							match children[partial.at(slice.len()) as usize] {
+								Some(x) => {
+									partial = partial.mid(slice.len() + 1);
+									key_nibbles += slice.len() + 1;
+									x
+								},
+								None => {
+									self.record(|| TrieAccess::NonExisting { full_key });
+
+									return Ok(None)
+								},
+							}
+						}
+					},
+					Node::Empty => {
+						self.record(|| TrieAccess::NonExisting { full_key });
+
+						return Ok(Some(hash))
+					},
+				};
+
+				// check if new node data is inline or hash.
+				match next_node {
+					NodeHandle::Hash(data) => {
+						hash = decode_hash::<L::Hash>(data)
+							.ok_or_else(|| Box::new(TrieError::InvalidHash(hash, data.to_vec())))?;
+						break
+					},
+					NodeHandle::Inline(data) => {
+						node_data = data;
```

Comment:
I am not sure this would work (I think we will loop on querying the same node again).
What I would do with inline node:
- not fetch new node but still descend in node
- return hash(node_data) if we are in an inline node instead of hash.

Reply:
IIUC for the `inline` case we could do something similar to:

```rust
NodeHandle::Inline(data) => {
    // Decode to ensure we are not following an empty node
    let decoded = match L::Codec::decode(data) {
	    Ok(node) => node,
	    Err(e) => return Err(Box::new(TrieError::DecoderError(hash, e))),
    };
    return match decoded {
	    Node::Empty => Ok(None),
	    _ => Ok(Some(hash)),
    }
},
```

With this change applied, the `test_merkle_value` seems to fail since we inlined

```
(b"AABBBB".to_vec(), vec![4; 1]),
(b"AB".to_vec(), vec![5; 1]),
(b"B".to_vec(), vec![6; 1]),
```

The failure is happening for key `let hash = trie.get_closest_merkle_value(b"AABC").unwrap();` which should not be present. Let me know if I got this right, thanks!